qgreenland/util/config.py

```diff
@@ -94,15 +99,120 @@ def _deref_boundaries(cfg: Dict[str, Any]) -> None:
         }
 
 
+class PartialFormatDict(dict):
+    """Hack to enable partial formatting of strings with {slugs}.
+
+    e.g.:
+
+        '{foo} {bar}, {baz}'.format_map(
+            PartialFormatDict(foo='Hello', bar='there')
+        )
+        >>> 'Hello there, {baz}'
+    """
+
+    def __missing__(self, key):
+        return '{' + key + '}'
+
+
+DictOrList = TypeVar(
+    'DictOrList',
+    Dict[Any, Any],
+    Tuple[Any],
+    List[Any],
+)
+
+
+def _interpolate_nested_values(
+    thing: DictOrList,
+    **kwargs,
+) -> DictOrList:
+    """Interpolate all strings found in `thing` with `kwargs`.
+
+    WARNING: Mutates `thing`!
+    """
+    items: Iterable[Any]
+    if isinstance(thing, dict):
+        items = thing.items()
+    elif isinstance(thing, list):
+        items = enumerate(thing)
+    elif isinstance(thing, str):
+        return thing.format_map(PartialFormatDict(**kwargs))
+    else:
+        return thing
+
+    for key, value in items:
+        # 🌶️ Recurse 🌶️!
+        thing[key] = _interpolate_nested_values(value, **kwargs)
+
+    return thing
+
+
+def _interpolate_template_kwargs(
+    template: Dict[str, Any],
+    **kwargs: Dict[str, Any],
+) -> List[Dict[str, Any]]:
+    # TODO: kwarg default values (i.e. optional kwargs) would be cool to
+    # provide. Careful, cowboy, we're getting into Jinja-like templating
```

Comment:
Maybe we should use jinja for this job? It may expose some functionality that we don't need right now, but it is well known and widely used.

:thinking: maybe it could even be used to clean up redundancies in e.g., the wmm layers?